### ISSUE TITLE
s6-linux-init: fatal: unable to copy /run/s6/basedir/run-image to /run: Operation not permitted

### ISSUE THREAD

**Issue by andersnev** [just-containers/s6-overlay#526]
Hi, 
I'm trying to run s6-overlay in a kubernetes environment with a read only filesystem and a non root user (as well as other security policies).

I suspect one of the security policies is preventing me from running s6, so I was wondering if you could help me figure out what permissions I would need to fix this error: 

```
s6-overlay-suexec: warning: unable to gain root privileges (is the suid bit set?)
/package/admin/s6-overlay/libexec/preinit: info: read-only root
/package/admin/s6-overlay/libexec/preinit: info: writeable /run. Checking for executability.
s6-linux-init: fatal: unable to copy /run/s6/basedir/run-image to /run: Operation not permitted
```

I've set S6_READ_ONLY_ROOT=1, dropped privileges using the USER directive in the Dockerfile and mounted /run as a tmpfs in the container.  

Dockerfile:
```
RUN useradd -u 10001 user && \
    mkdir -p /var/run/s6/ && \
    chown -c 10001 --recursive /var/run/s6/ && \
    chown -c 10001 --recursive /run/


USER user

ENTRYPOINT ["/init"]
```

Kubernetes container and security context: 
```
      volumes:
        - name: tmp
          emptyDir: {}
        - name: run
          emptyDir:
            medium: "Memory"
      containers:
      - name: app-name
        image: image-name
        env:
          - name: "S6_READ_ONLY_ROOT"
            value: "1"
        securityContext:
          allowPrivilegeEscalation: false
          privileged: false
          readOnlyRootFilesystem: true
          runAsNonRoot: true
          runAsUser: 10001
        volumeMounts:
          - name: "tmp"
            mountPath: /tmp
          - name: "run"
            mountPath: /run
```

**Comment by skarnet** [comment]
Hmmm, that's interesting. Would it be possible for you to also set the *group* for `/run`, and to make sure `/run` has permissions `02755`? (`drwxr-sr-x`) It is possible that the copy to `/run` fails because of that.

If that doesn't change anything, it may be a bug in the library call that copies `run-image` recursively; in order to be sure I would need to see a strace of s6-linux-init. To do this, please:
- install the `strace` package on your image
- make sure you log the console output (it can be very verbose)
- when you install the overlay, in `/package/admin/s6-overlay/libexec/stage0`, replace the last line with: `exec strace "$basedir/bin/init" "$@"`
- make sure you log the console output (it can be very verbose)

Towards the end there will be a line ending with `-1 EPERM`, that's the one I need (with a little context around it), and I'll be able to tell you whether it's something in your configuration or a bug I need to fix.

**Comment by andersnev** [comment]
Thanks for the quick reply! I added a group and set /run/ permissions to 02755, but it still gives the same error.
Dockerfile:
```
RUN groupadd -g 10001 s6_group && \ 
    useradd -u 10001 s6_user && \
    mkdir -p /var/run/s6/ && \
    chown 10001:10001 --recursive /var/run/s6/ && \
    chown 10001:10001 --recursive /run/ && \
    chmod 02755 /run/

USER s6_user:s6_group

ENTRYPOINT ["/init"]
```
I tried tracing the console output like you said, but I couldn't figure out how to run `strace` as a non root user in this environment. There are security policies in place preventing me from escalating privilege for the container and adding `SYS_PTRACE` capabilities, which it looks like `strace` needs. 

```
/package/admin/s6-overlay/libexec/preinit: info: read-only root
/package/admin/s6-overlay/libexec/preinit: info: writable /run. Checking for executability.
strace: test_ptrace_get_syscall_info: PTRACE_TRACEME: Operation not permitted
strace: ptrace(PTRACE_TRACEME, ...): Operation not permitted
strace: PTRACE_SETOPTIONS: Operation not permitted
strace: detach: waitpid(21): No child processes
strace: Process 21 detached
```

**Comment by skarnet** [comment]
Can't you temporarily bypass the security policy for debugging purposes? I'm sorry, but I cannot help with this without an strace output. I tried to reproduce the issue, but it works on my setup, so the problem lies in the interaction between your constraints and what s6-overlay does, and strace is by far the fastest way to pinpoint it.

**Comment by skarnet** [comment]
Ah, obviously, my bad - when you strace init, it doesn't run as pid 1 anymore, so the behaviour changes. We need a workaround. Please try something like this:

```
ADD --chown=root:root --chmod=0755 init-strace /init-strace

# replaces ENTRYPOINT ["/init"] for debugging purposes
ENTRYPOINT ["/init-strace"] 
```

and the contents of your local `init-strace` file should be, unless I've made a mistake:

```sh
#!/bin/sh -e
# change the path depending on where your distribution installs the strace binary
/usr/bin/strace -p 1 </dev/null >/run/strace-output 2>&1 3>&- &
exec /init "$@"
```

The trace will then be available in `/run/strace-output`.

Thanks!

**Comment by filippog** [comment]
I ran into this problem too, and with the `strace` set up as `init-strace` I get the following:

```
485566 lchown("/run/service/.s6-svscan/SIGQUIT", 1000, 1000) = 0
485566 chmod("/run/service/.s6-svscan/SIGQUIT", 0100755) = 0 
485566 chmod("/run/service/.s6-svscan", 040755) = 0
485566 lchown("/run/service/.s6-svscan", 1000, 1000) = 0
485566 chmod("/run/service/.s6-svscan", 040755) = 0
485566 chmod("/run/service", 040755)    = 0
485566 lchown("/run/service", 1000, 1000) = 0
485566 chmod("/run/service", 040755)    = 0
485566 chmod("/run", 040755)            = -1 EPERM (Operation not permitted)
485566 writev(2, [{iov_base="s6-linux-init: fatal: unable to copy /run/s6/basedir/run-image to /run: Operation not permitted\n", iov_len=96}, {iov_base=NULL, iov_len=0}], 2) = 96
```

hope that helps!

**Comment by skarnet** [comment]
It's fixed in the current git of the underlying software, but I have another project to finish before I tag numbered releases (just in case I need to add another functionality).

If you're willing to build s6-overlay yourself with the latest git commits of skalibs and s6-linux-init, you will get the fixed version. Otherwise, please be patient, I'll tag a numbered release as soon as possible.

**Comment by andersnev** [comment]
Great! I will give that a go, thanks a lot

**Comment by sambercovici** [comment]
TLDR - I just compiled from source and it worked!
When do you plan to release the next version?

Compiled on ubuntu 18.04 targeting default musl (for an alpine container)
1) commenting in conf/versions SKALIBS_VERSION and S6_LINUX_INIT_VERSION so that make will use latest.
2) needed to add to mk/skaware.mk  for SKALIBS_OPTIONS --with-sysdep-posixspawnearlyreturn=no
3) Manually deleted in output/sources/skalibs/Makefile 
$(shell test -r config.mak)
ifneq ($(.SHELLSTATUS),0)
$(error Missing config.mak; please use ./configure first)
endif

**Comment by skarnet** [comment]
The next version will happen at the same time as a massive skarnet.org release, that's planned for this summer when everything's ready - I'm working on it but it could still be one month or two.

**Comment by sambercovici** [comment]
> The next version will happen at the same time as a massive skarnet.org release, that's planned for this summer when everything's ready - I'm working on it but it could still be one month or two.

Do you have a clearer date for this release?

**Comment by skarnet** [comment]
Not yet, sorry. Probably early to mid September.

**Comment by sambercovici** [comment]
HI Again, is the release still planned for mid September?

**Comment by skarnet** [comment]
It will probably be delayed a little (I have a lot of work to do before it's ready), but it shouldn't be very long. Before the end of the month, I hope.

**Comment by rauanmayemir** [comment]
We’re hitting this issue, too.

Not only can’t I run non-privileged containers with read only root fs, I also can’t specify s6-setuidguid because after starting the container, s6 then tries to drop privileges and start a logutil service. (but since the user is already non-privileged, it can’t change privileges)

**Comment by skarnet** [comment]
Sorry for the delay, the release is going to come out as soon as it's ready.

**Comment by rauanmayemir** [comment]
Reporting that it IS fixed! 😌

**Comment by skarnet** [comment]
Awesome!
